// postgraphql/build/postgraphql/schema/procedures/createPgProcedureMutationGqlFieldEntry.js
"use strict";
var tslib_1 = require("tslib");
var pluralize = require("pluralize");
var graphql_1 = require("graphql");
var interface_1 = require("../../../interface");
var utils_1 = require("../../../graphql/utils");
var createMutationGqlField_1 = require("../../../graphql/schema/createMutationGqlField");
var createCollectionRelationTailGqlFieldEntries_1 = require("../../../graphql/schema/collection/createCollectionRelationTailGqlFieldEntries");
var utils_2 = require("../../../postgres/utils");
var PgCollection_1 = require("../../../postgres/inventory/collection/PgCollection");
var pgClientFromContext_1 = require("../../../postgres/inventory/pgClientFromContext");
var createPgProcedureFixtures_1 = require("./createPgProcedureFixtures");
var createPgProcedureSqlCall_1 = require("./createPgProcedureSqlCall");
var createConnectionGqlField_1 = require("../../../graphql/schema/connection/createConnectionGqlField");
/**
 * Creates a single mutation GraphQL field entry for our procedure. We use the
 * `createMutationGqlField` utility from the `graphql` package to do so.
 */
// TODO: test
function createPgProcedureMutationGqlFieldEntry(buildToken, pgCatalog, pgProcedure) {
    var inventory = buildToken.inventory;
    var fixtures = createPgProcedureFixtures_1.default(buildToken, pgCatalog, pgProcedure);
    // See if the output type of this procedure is a single object, try to find a
    // `PgCollection` which has the same type. If it exists we add some extra
    // stuffs.
    var pgCollection = !pgProcedure.returnsSet
        ? inventory.getCollections().find(function (collection) { return collection instanceof PgCollection_1.default && collection.pgClass.typeId === fixtures.return.pgType.id; })
        : null;
    // Create our GraphQL input fields users will use to input data into our
    // procedure.
    var inputFields = fixtures.args.map(function (_a) {
        var name = _a.name, gqlType = _a.gqlType;
        return [utils_1.formatName.field(name), {
                // No description…
                type: pgProcedure.isStrict ? new graphql_1.GraphQLNonNull(graphql_1.getNullableType(gqlType)) : gqlType,
            }];
    });
    return [utils_1.formatName.field(pgProcedure.name), createMutationGqlField_1.default(buildToken, {
            name: pgProcedure.name,
            description: pgProcedure.description,
            inputFields: inputFields,
            outputFields: [
                [utils_1.formatName.field(pgProcedure.returnsSet
                        ? pluralize(getTypeFieldName(fixtures.return.type))
                        : getTypeFieldName(fixtures.return.type)), {
                        // If we are returning a set, we should wrap our type in a GraphQL
                        // list.
                        type: pgProcedure.returnsSet
                            ? new graphql_1.GraphQLList(fixtures.return.gqlType)
                            : fixtures.return.gqlType,
                        resolve: function (value) { return fixtures.return.intoGqlOutput(value); },
                    }],
                // An edge variant of the created value. Because we use cursor
                // based pagination, it is also helpful to get the cursor for the
                // value we just created (thus why this is in the form of an edge).
                // Also Relay 1 requires us to return the edge.
                //
                // We may deprecate this in the future if Relay 2 doesn’t need it.
                pgCollection && pgCollection.paginator && [utils_1.formatName.field(pgCollection.type.name + "-edge"), {
                        description: "An edge for the type. May be used by Relay 1.",
                        type: createConnectionGqlField_1.getEdgeGqlType(buildToken, pgCollection.paginator),
                        args: { orderBy: createConnectionGqlField_1.createOrderByGqlArg(buildToken, pgCollection.paginator) },
                        resolve: function (value, args) { return ({
                            paginator: pgCollection.paginator,
                            ordering: args['orderBy'],
                            cursor: null,
                            value: value,
                        }); },
                    }]
            ].concat((pgCollection ? createCollectionRelationTailGqlFieldEntries_1.default(buildToken, pgCollection, { getCollectionValue: function (value) { return value; } }) : [])),
            // Actually execute the procedure here.
            execute: function (context, gqlInput) {
                return tslib_1.__awaiter(this, void 0, void 0, function () {
                    var client, input, procedureCall, aliasIdentifier, query, rows, values, _a, _b;
                    return tslib_1.__generator(this, function (_c) {
                        switch (_c.label) {
                            case 0:
                                client = pgClientFromContext_1.default(context);
                                input = inputFields.map(function (_a, i) {
                                    var fieldName = _a[0];
                                    return fixtures.args[i].fromGqlInput(gqlInput[fieldName]);
                                });
                                procedureCall = createPgProcedureSqlCall_1.default(fixtures, input);
                                aliasIdentifier = Symbol();
                                query = utils_2.sql.compile(
                                // If the procedure returns a set, we must select a set of values.
                                pgProcedure.returnsSet
                                    ? (_a = ["select to_json(", ") as value from ", " as ", ""], _a.raw = ["select to_json(", ") as value from ", " as ", ""], utils_2.sql.query(_a, utils_2.sql.identifier(aliasIdentifier), procedureCall, utils_2.sql.identifier(aliasIdentifier))) : (_b = ["select to_json(", ") as value"], _b.raw = ["select to_json(", ") as value"], utils_2.sql.query(_b, procedureCall)));
                                return [4 /*yield*/, client.query(query)];
                            case 1:
                                rows = (_c.sent()).rows;
                                values = rows.map(function (_a) {
                                    var value = _a.value;
                                    return fixtures.return.type.transformPgValueIntoValue(value);
                                });
                                // If we selected a set of values, return the full set. Otherwise only
                                // return the one we queried.
                                return [2 /*return*/, pgProcedure.returnsSet ? values : values[0]];
                        }
                    });
                });
            },
        })];
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = createPgProcedureMutationGqlFieldEntry;
/**
 * Gets the field name for any given type. Pluralizes the name of for item
 * types in list types.
 */
function getTypeFieldName(_type) {
    return interface_1.switchType(_type, {
        nullable: function (type) { return getTypeFieldName(type.nonNullType); },
        list: function (type) { return pluralize(getTypeFieldName(type.itemType)); },
        alias: function (type) { return type.name; },
        enum: function (type) { return type.name; },
        object: function (type) { return type.name; },
        scalar: function (type) { return type.name; },
    });
}
